### PR TITLE
S175 : move in mapreduce

### PR DESCRIPTION
**Pull request by eschultink**
copied all the MR stuff over into pipelines project; don't need it separate.

### Change implications

 - breaking change to API? **no**
 - changes dependencies?  **yes**

### CODE REVIEW

**Review by jlorper — APPROVED** [review]

**Review by aperez-worklytics — APPROVED** [review]

File: java/example/war/lettercount.jsp [hunk 1]
```diff
@@ -33,7 +36,7 @@
     String text = request.getParameter(TEXT_PARAM_NAME);
     String pipelineId = request.getParameter(PIPELINE_ID_PARAM_NAME);
     String cleanupId = request.getParameter(CLEANUP_PIPELINE_ID_PARAM_NAME);
-    PipelineService service = PipelineServiceFactory.newPipelineService();
+   PipelineService service = new PipelineServiceImpl(new PipelineManager(AppEngineBackEnd.))
```
Comment (jlorper) [inline java/example/war/lettercount.jsp:39]: I see it is the example, but that won't compile

File: java/pom.xml [hunk 2]
```diff
@@ -141,10 +154,25 @@
...
+      <version>1.3</version>
+    </dependency>
+    <dependency>
+      <groupId>org.json</groupId>
+      <artifactId>json</artifactId>
+      <version>20231013</version>
+    </dependency>
+    <dependency>
+      <groupId>it.unimi.dsi</groupId>
+      <artifactId>fastutil</artifactId>
```
Comment (jlorper) [inline java/pom.xml:169]: what special stuff do we need for collections?

File: docs/design-patterns.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+2. make all classes dependent on `Datastore` not singletons; and take it as a constructor parameter
+
+Current choice is (2), because limits scope of changes for users; will just have to change how they
+get `PipelineService`/`PipelineManager` instances for launching jobs; and initially, for case of 
+running everything within default project/database/namespace, it can remain a singleton for them.
+
+
+## What is the right pattern long-term?
+
+Passing options around allows classes to be self-evidently threadsafe; no risk that one thread
```
Comment (jlorper) [inline docs/design-patterns.md:17]: everything happens in request/response model, right? Can we pass through something (a header, a param) that identifies the org and provide a Database instance in a filter, available for the whole request?

File: java/src/main/java/com/google/appengine/tools/mapreduce/impl/util/RequestUtils.java [hunk 4]
```diff
@@ -1,23 +1,59 @@
...
+
+/**
+ * handles translation to/from request parameters and pipeline backends
+ *
+ * q: good pattern? kinda franken factory to build instances from requests ... but w/o this, gets repeated in a bunch
+ *  of places
+ */
+@Singleton
+@AllArgsConstructor(onConstructor_ = @Inject)
 public class RequestUtils {
```
Comment (jlorper) [inline java/src/main/java/com/google/appengine/tools/mapreduce/impl/util/RequestUtils.java:26]: This, this should be a web Filter, create the datastore instance in the filter and pass it through a request.setAttribute, then available everywhere.

**Review by jlorper — COMMENTED** [review]
Some comments, still stuff to review

**Pull request merged**